### PR TITLE
ENH: Pass optimizations arguments to asv build

### PR DESCRIPTION
**Pull request by seiko2plus**
####  ENH: Pass optimizations arguments to asv build #17284 
  
 This patch allows passing `-j`, `--cpu-baseline`, `--cpu-dispatch` and `--disable-optimization`
  to ASV build when argument `--bench-compare` is used.

TODO:
- [x] local test

### CODE REVIEW

**Comment by seiko2plus** [comment]
@seberg, I just don't understand why ASV doesn't provide a direct way to specify custom build options?  Am I missing something?

**Comment by seberg** [comment]
I looked around a bit, and don't see a way other than as part of the config. However, we could use `--config` to use a custom config file rather than deleting the current config? May want to ping Pauli, though, at least if that doesn't see right.

**Comment by Qiyu8** [comment]
I think this is a good solution to benchmark flexible universal intrinsic usage. but asv community still needs to considering the necessity of adding `{build_args}` to `asv.config.json`. @pv what do you think?

**Comment by seiko2plus** [comment]
@seberg,
> I looked around a bit, and don't see a way other than as part of the config. 

I confirm it, thank you!

> we could use --config to use a custom config file rather than deleting the current config?

I followed your suggestion, I kept the main config file as-is so it can be used with `ASV` command.

@Qiyu8,
> I think this is a good solution to benchmark flexible universal intrinsic usage.

An effective workaround but we still have other obstacles such as providing auto-generated test data and
provides the ability to check the sanity of it the same as [OpenCV](https://github.com/opencv/opencv/wiki/HowToUsePerfTests) do, another nightmare I guess.
It would be easier for me to drop `ASV` and create our own benchmark tool, but I guess the community wouldn't easily
accept such a move like that.

File: runtests.py [hunk 1]
```diff
@@ -279,7 +283,7 @@ def main(argv):
             bench_args.extend(['--bench', a])
 
         if not args.bench_compare:
-            cmd = ['asv', 'run', '-n', '-e', '--python=same'] + bench_args
+            cmd = ['asv', 'run', '-n', '-e', '--python=same'] + asv_cfpath + bench_args
```
Comment (seiko2plus) [inline runtests.py:286]:
```suggestion
            cmd = ['asv', 'run', '-n', '-e', '--python=same'] + bench_args
```
we only need custom build options for bench_compare

**Comment by seberg** [comment]
Maybe its right and we should consider upstreaming things to ASV.  I am a bit worried that running plain ASV won't have access to this, but there may not be a good way to do that.
Does this give the baseline/optimized/disabled versions different environment names in the ASV run?

File: runtests.py (hunk 1 above)
Comment (seiko2plus) [reply runtests.py:286]: Yes indeed,  where is my mind?

**Comment by seiko2plus** [comment]
@seberg,
>  Does this give the baseline/optimized/disabled versions different environment names in the ASV run?

No, it generates a config file with the required build options then pass the generated config file through
`--config`, it also clears the whole cache if the user changed the build options.

File: runtests.py [hunk 2]
```diff
@@ -491,6 +494,96 @@ def build_project(args):
...
+    if args.cpu_dispatch:
+        build += ["--cpu-dispatch", args.cpu_dispatch]
+    if args.disable_optimization:
+        build += ["--disable-optimization"]
+
+    is_cached = asv_substitute_config(conf_path, nconf_path,
+        numpy_build_options = ' '.join([f'\\"{v}\\"' for v in build]),
+    )
+    if not is_cached:
+        asv_clear_cache(bench_path)
```
Comment (seiko2plus) [inline runtests.py:516]: only clear the cache for the involved heads

**Comment by seiko2plus** [comment]
@seberg,
> I am a bit worried that running plain ASV won't have access to this, but there may not be a good way to do that.

I didn't touch the original config file,  I created a new config header instead `asv_compare.conf.json` and it only works through `--bench-compare`.

File: benchmarks/asv_compare.conf.json [hunk 3]
```diff
@@ -0,0 +1,95 @@
+// This config file is almost similar to 'asv.conf.json' except it contains
```
Comment (seiko2plus) [inline benchmarks/asv_compare.conf.json:1]: rename `asv_compare.conf.json` to `asv_compare.conf.json.tpl`

File: runtests.py [hunk 4]
```diff
@@ -491,6 +498,97 @@ def build_project(args):
...
+    )
+    if not is_cached:
+        asv_clear_cache(bench_path, h_commits)
+    return nconf_path
+
+def asv_clear_cache(bench_path, h_commits, env_dir="env"):
+    """
+    Force ASV to clear the cache according to specified 'h_commits'
+    """
+    asv_build_pattern = os.path.join(bench_path, env_dir, "*", "asv-build-cache")
```
Comment (seiko2plus) [inline runtests.py:531]:
```suggestion
    asv_build_pattern = os.path.join(bench_path, env_dir, env_hash, "asv-build-cache")
```

@seberg, the only downside is here, `asv_clear_cache` clear involved commits cache dir from all environment paths

Comment (seiko2plus) [reply runtests.py:532]: not a big issue I guess.

**Review by seiko2plus — COMMENTED** [review]
Looks good to me, what do you think @pv?

**Comment by charris** [comment]
This needs documentation so people will know how to use it. Maybe in the `README.rst` or `runtests.py` files?

**Comment by charris** [comment]
Thanks Sayed. I put this in to aid the ongoing SIMD work, but more usage documentation would be helpful.

**Pull request merged**